ccdb/law/views.py

```diff
@@ -53,107 +55,139 @@ def get(self, request):
         return render(request, 'law/feedback.html', dict())
 
 
-@user_passes_test(lambda u: u.is_staff)
-def edit_index(request):
-    return render(request, 'law/edit_index.html',
-                  dict(snapshots=Snapshot.objects.all(),
-                       working_snapshot=working_snapshot(),
-                       public_snapshot=public_snapshot()))
+class StaffMixin(object):
+    @method_decorator(login_required)
+    @method_decorator(user_passes_test(lambda u: u.is_staff))
+    def dispatch(self, *args, **kwargs):
+        return super(StaffMixin, self).dispatch(*args, **kwargs)
 
 
-@user_passes_test(lambda u: u.is_staff)
-def edit_snapshots(request):
-    return render(request, 'law/edit_snapshots_index.html',
-                  dict(snapshots=Snapshot.objects.all().order_by("-created")))
+class WorkingSnapshotMixin(object):
+    def snapshot(self):
+        return working_snapshot()
 
 
-@user_passes_test(lambda u: u.is_staff)
-def edit_snapshot(request, id):
-    return render(request, 'law/edit_snapshot.html',
-                  dict(snapshot=get_object_or_404(Snapshot, id=id)))
+class PublicSnapshotMixin(object):
+    def snapshot(self):
+        return public_snapshot()
 
 
-@user_passes_test(lambda u: u.is_staff)
-def clone_snapshot(request, id):
-    snapshot = get_object_or_404(Snapshot, id=id)
-    snapshot.clone(label=request.POST.get('label', 'new snapshot'),
-                   user=request.user,
-                   description=request.POST.get('description', ''))
-    return HttpResponseRedirect("/edit/")
+class ChargeLocatorMixin(object):
+    def get_charge_by_slugs(self, slugs):
+        return self.snapshot().get_charge_by_slugs(slugs)
```

Comment:
Adding the get_charge_by_slugs to the mixin requires a View to inherit from a *SnapshotMixin first, which seems a little mysterious. The prior approach -- calling the snapshot model's get_charge_by_slugs(slugs) method directly seems simpler.